tests/functions/test_visualizations.py

```diff
@@ -264,3 +263,309 @@ def test_clustergrammerify_single_sample(metadata_df):
+            "Participant Id: o",
+            "Cohort: i",
+            "Sex: j",
+            "OS: e",
+            "Died: False",
+        ),
+        (
+            "Sample Id: CT5",
+            "Participant Id: a",
+            "Cohort: c",
+            "Sex: d",
+            "OS: e",
+            "Died: False",
+        ),
+        (
+            "Sample Id: CT6",
+            "Participant Id: g",
+            "Cohort: i",
+            "Sex: j",
+            "OS: k",
+            "Died: True",
+        ),
+        (
+            "Sample Id: CT7",
+            "Participant Id: m",
+            "Cohort: p",
+            "Sex: d",
+            "OS: n",
+            "Died: True",
+        ),
+        (
+            "Sample Id: CT8",
+            "Participant Id: o",
+            "Cohort: i",
+            "Sex: j",
+            "OS: e",
+            "Died: False",
+        ),
+    ]
+    categories = _metadata_to_categories(md_toomany)
+    assert cat_toomany == categories
+
+    md_alldiff = md_names.copy()
+    md_alldiff.index = ["CT5", "CT6", "CT7", "CT8"]
+    md_alldiff = pd.concat([md_names, md_alldiff], axis=0)
+    md_alldiff["capital_test"] = range(8)  # dropped because cardinality == md.shape[0]
+    cat_alldiff = [
+        (
+            "Sample Id: CT1",
+            "Participant Id: a",
+            "Cohort: c",
+            "Sex: d",
+            "OS: e",
+            "Died: False",
+        ),
+        (
+            "Sample Id: CT2",
+            "Participant Id: g",
+            "Cohort: i",
+            "Sex: j",
+            "OS: k",
+            "Died: True",
+        ),
+        (
+            "Sample Id: CT3",
+            "Participant Id: m",
+            "Cohort: p",
+            "Sex: d",
+            "OS: n",
+            "Died: True",
+        ),
+        (
+            "Sample Id: CT4",
+            "Participant Id: o",
+            "Cohort: i",
+            "Sex: j",
+            "OS: e",
+            "Died: False",
+        ),
+        (
+            "Sample Id: CT5",
+            "Participant Id: a",
+            "Cohort: c",
+            "Sex: d",
+            "OS: e",
+            "Died: False",
+        ),
+        (
+            "Sample Id: CT6",
+            "Participant Id: g",
+            "Cohort: i",
+            "Sex: j",
+            "OS: k",
+            "Died: True",
+        ),
+        (
+            "Sample Id: CT7",
+            "Participant Id: m",
+            "Cohort: p",
+            "Sex: d",
+            "OS: n",
+            "Died: True",
+        ),
+        (
+            "Sample Id: CT8",
+            "Participant Id: o",
+            "Cohort: i",
+            "Sex: j",
+            "OS: e",
+            "Died: False",
+        ),
+    ]
+    categories = _metadata_to_categories(md_alldiff)
+    assert cat_alldiff == categories
+
+    # others test lack of combination with <= 3 columns after filtering
+    # others test always keep 'Cohort' and 'Collection Event'
```

Comment:
These sound worth adding before merging.

Reply:
oh sorry, this is referring to the fact that the other already existing tests would fail if these conditions weren't handled properly
clarifying comments